### DISCUSSION TITLE
Managing Security Onion log retention

### DISCUSSION THREAD

**Discussion by DrewHorrigan** [Security-Onion-Solutions/securityonion#7290] [Unsupported Versions] [answered]
Hello, I've been having trouble finding information on what should be basic housekeeping functionality - specifically, managing the distribution of disk space between Elasticsearch database, raw logs and pcaps.

Our current use case is that Zeek logs take up the majority of disk space when the node is left to its own devices, which leaves very little space for pcap data. I have been looking all over the internet for "proper" ways to manage Zeek log expiration, but everything I found was lacking in some way:

- **so-sensor-clean** - CRIT_DISK_USAGE is hardcoded, and thus non-configurable; the script also cleans all logs and pcaps nondiscriminately. My understanding is that it's a last resort to make sure the disk does not fill up all the way.
- **log_size_limit setting** - [this was apparently the go-to solution](https://groups.google.com/g/security-onion/c/n2vA2cqBWqU) in the old version, but from what I can see from Salt pillar files in version 2, this setting only affects the Elasticsearch database (I can't tell for sure, as it's not documented).
- Pillar values **zeek.zeekctl.LogRotationInterval** & **zeek.zeekctl.LogExpireInterval** - potentially solve my problem by limiting the size of Zeek logs, but, despite being mentioned in the manual, these values are only used when running 'zeekctl cron' from so-zeek container, which is not scheduled anywhere.

With all that in mind, what is the intended way to manage the allotted disk space / retention interval for different file types, or at least specifically Zeek logs?

**Comment by dougburks** [comment, marked as answer]
Here's the documentation for `log_size_limit`:
https://docs.securityonion.net/en/2.3/elasticsearch.html?highlight=log_size_limit#deleting-indices

What kind of deployment do you have (eval, standalone, distributed)?

What is the output of the following?
```
df -h
```
```
sudo du --max-depth=1 -h /nsm
```

**Reply by Acewiza** [reply]
I suspect nsm disk space epidemiology is highly dependent on the traffic environment.  Pcap is by far the biggest consumer in my standalone case.

**Reply by DrewHorrigan** [reply]
> What kind of deployment do you have (eval, standalone, distributed)?

We use a distributed deployment with 1 management node and 2 heavy nodes (yes, I know, it's not recommended). The problem is most pronounced on the second node, which sees more traffic (at least since last time I purged Zeek logs using `zeekctl cron`).

> What is the output of the following?

Heavy node 1:
```
[username@heavynode1 ~]$ df -h
Filesystem                     Size  Used Avail Use% Mounted on
devtmpfs                       158G     0  158G   0% /dev
tmpfs                          158G   80K  158G   1% /dev/shm
tmpfs                          158G  148M  158G   1% /run
tmpfs                          158G     0  158G   0% /sys/fs/cgroup
/dev/mapper/system-root        293G   34G  259G  12% /
/dev/mapper/system-tmp         2.0G   33M  2.0G   2% /tmp
/dev/sda2                      497M  190M  307M  39% /boot
/dev/mapper/vg_app-lv_app_nsm  2.0T  1.8T  245G  89% /nsm
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/5b820c05f44986ede8b5c751480691ee99e83ec07e95c95f283de744c7858b8d/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/11acd86c79a1c2cc86c10e89d317dea6afbefcd38961cdac2dc88e3040656204/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/2f3b09bd15b2324c353c74cc91e0ed158cd22f141a1ef1c119547953cf33f215/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/3c5d39a132ec436efdec792a6539382b36422c548522bdc56c1c84aaaf620387/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/81a3bc89fc1372ea2bc94019888a725b3153dbe7eafe16f38d442369314d52c8/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/61ffc42da5c81f4e930745e45213036d5ef16fa42a89b5afa210f7d8e0b637bf/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/88a8a3498e8dee6f9f6ed6a1b81c909d8c3f72d38590c891fb29243489480338/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/23cf8f7d480fd6ff9c420e0b1030ee30b54426bb135be7c6f528ca8c56401f41/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/b9b23b778a924729707e8c512161edffd52fdcf16241df88515b5aeb110c1152/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/b515da17a7ec4698652c22b440bb6b293c655d794c0ed39ba69a9abd40496306/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/c343bf487770ee46e74985f982b4ff98b44fba9e261a3dc246c2926c923f2014/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/f53b91370f5c9c096b4ddcb25b5c4c263329a0a168d4d3a668e24d444929f851/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/a9a7cd2d68faa5ecfd76e406834b48486f478bcf6041fcfac4cd194f892f8ffa/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/bd151f891bcd78f9d68c12e064ea93bf9c98a25a4ead7365a268587e8aa5d700/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/4ac503df4d60a8a97bba6d33d82bb737dcb7732247949a80e75179e5471fedf2/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/d0428e5a08f04bcc42e57cda17d93742f2e3a6b273e7fb65144d50bcaeab568a/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/f224361651187dcd7e19377be24d22724db78f293fd88504372fe628f6c922ff/merged
overlay                        293G   34G  259G  12% /var/lib/docker/overlay2/9697f6cbc8c6257566d0c75f573c7237ac3c907b4fd850eac33453d6ca896cf4/merged
tmpfs                           32G     0   32G   0% /run/user/0
tmpfs                           32G     0   32G   0% /run/user/1000
[username@heavynode1 ~]$ sudo du --max-depth=1 -h /nsm
[sudo] password for username:
5.9G    /nsm/docker-registry
698M    /nsm/repo
483G    /nsm/elasticsearch
1.2T    /nsm/pcap
24K     /nsm/pcaptmp
0       /nsm/pcapout
29G     /nsm/pcapindex
4.2G    /nsm/suricata
0       /nsm/import
4.0K    /nsm/logstash
11G     /nsm/strelka
131G    /nsm/zeek
1013M   /nsm/wazuh
1.8T    /nsm
```

Heavy node 2:
```
[username@heavynode2 ~]$ df -h
Filesystem                     Size  Used Avail Use% Mounted on
devtmpfs                       158G     0  158G   0% /dev
tmpfs                          158G   80K  158G   1% /dev/shm
tmpfs                          158G  3.9G  154G   3% /run
tmpfs                          158G     0  158G   0% /sys/fs/cgroup
/dev/mapper/system-root        293G   32G  262G  11% /
/dev/mapper/system-tmp         2.0G   33M  2.0G   2% /tmp
/dev/sda2                      497M  190M  307M  39% /boot
/dev/mapper/vg_app-lv_app_nsm  2.0T  1.8T  226G  89% /nsm
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/e512c523c40c3d272d06456c0b00bcc5c876be75c0b0794c26f4cd4c1d92c7c8/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/bf8636e64d9ac00d463fc7e4cd1815a88020c6faf00696736e63e6c02a7bb78d/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/437a9997b9d3ee83e2067fdd6ede7d9136487e9dab9f2f91228ab15df989f7a4/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/b9b0cd81221c952e61ee7c8f02327ea79fd27f7989ca15466791c17f698b0a88/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/213694902c8585930fab9a4e89d40c4767dbd21aee500fd5809a1bb5e3ddeef7/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/92d2b85d32832cb75f90734cf6ee8607122c6a376a18d41fc5b90b2525434df6/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/ab46f397411b5234cd685973a00f6aeb64cd1fd97cbf26091056eaa54b334996/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/0f46d8da51735ec6d87d187dd2c2de8726022e2bcc4d2df7c2dc7beca6cd99bc/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/fb97a430d47754742dd868dd1648129804873edfb64be3242e729e51bb174c4c/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/df64954cda761bebe2a84c8a55396d14a0362109d99069e1f5e3dfa6bd5660df/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/649489b4e3260e9f5fcfaea047d9d157d3ae5d9cf877161c0f37d64ab576f37d/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/bcdedb313591fc51f672a0f63fc670f35c681b720f1716ee7085ee2eeeee5b92/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/2331a941686369334109a86cf96a0822718e2a751cc8876c3c6a040476d8cf7f/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/1300cc6490e6de35a1294548691335826b3ff359670baa0e8a70fd869de1a74c/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/ea9f98d46bb2f4e6d3eb0f736b977fec89e7e5c67d513b8be88d796da83b45f0/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/1acbeb2343691014946df531bfadf24b5ee251fa468d734f6a86f94f1ad9ba90/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/e93ee26d166bb287b4f855b45cc99a2ea898d92fb752f644d57dc10dbbcf2a24/merged
overlay                        293G   32G  262G  11% /var/lib/docker/overlay2/5cbbc10056944dd77a6cdb9466a2db78b2d73e21051aa7a17d0766ae53c48de6/merged
tmpfs                           32G     0   32G   0% /run/user/1000
[username@heavynode2 ~]$ sudo du --max-depth=1 -h /nsm
[sudo] password for username:
5.9G    /nsm/docker-registry
698M    /nsm/repo
507G    /nsm/elasticsearch
18G     /nsm/pcap
20K     /nsm/pcaptmp
0       /nsm/pcapout
532M    /nsm/pcapindex
14G     /nsm/suricata
du: cannot access ‘/nsm/zeek/extracted/SMTP-FTyyrb4wKWR8v8btQf.pdf’: No such file or directory
1.1T    /nsm/zeek
0       /nsm/import
4.0K    /nsm/logstash
55M     /nsm/wazuh
148G    /nsm/strelka
1.7T    /nsm
```

**Reply by DrewHorrigan** [reply]
> Here's the documentation for `log_size_limit`: https://docs.securityonion.net/en/2.3/elasticsearch.html?highlight=log_size_limit#deleting-indices

My bad. In my defense, it doesn't show up in the internal search.

**Reply by dougburks** [reply]
> My bad. In my defense, it doesn't show up in the internal search.

It appears to show up for me at https://docs.securityonion.net/en/2.3/search.html?q=log_size_limit&check_keywords=yes&area=default:
![image](https://user-images.githubusercontent.com/1659467/156169627-0f688478-6b0e-4ae8-a353-6050f2511382.png)

> We use a distributed deployment with 1 management node and 2 heavy nodes (yes, I know, it's not recommended)

This is one of the reasons why we don't recommend heavy nodes. They experience resource contention between network services like Zeek and search services like Elasticsearch.

That being said, 1.1T of Zeek logs on heavynode2 seems unusual. 

Have you modified the Zeek configuration in any way? 

Have you configured it to extract additional files perhaps?

Have you looked inside /nsm/zeek/ to see if there is something in particular taking up more space than necessary? Perhaps some error log that is spinning out of control?

What is the output of the following?
```
sudo du --max-depth=1 -h /nsm/zeek
```

**Reply by DrewHorrigan** [reply]
> Have you modified the Zeek configuration in any way?
> 
> Have you configured it to extract additional files perhaps?

Not to my knowledge, no.



> What is the output of the following?

```
[username@heavynode2 ~]$ sudo du --max-depth=1 -h /nsm/zeek
[sudo] password for username:
900G    /nsm/zeek/logs
40G     /nsm/zeek/spool
2.2G    /nsm/zeek/extracted
942G    /nsm/zeek

[username@heavynode2 ~]$ sudo du --max-depth=1 -h /nsm/zeek/logs
57G     /nsm/zeek/logs/2022-03-02
57G     /nsm/zeek/logs/2022-03-03
56G     /nsm/zeek/logs/2022-02-18
41G     /nsm/zeek/logs/2022-02-19
41G     /nsm/zeek/logs/2022-02-20
57G     /nsm/zeek/logs/2022-02-21
58G     /nsm/zeek/logs/2022-02-22
57G     /nsm/zeek/logs/2022-02-23
59G     /nsm/zeek/logs/2022-02-24
54G     /nsm/zeek/logs/2022-02-25
39G     /nsm/zeek/logs/2022-02-26
38G     /nsm/zeek/logs/2022-02-27
56G     /nsm/zeek/logs/2022-02-28
58G     /nsm/zeek/logs/2022-03-01
236K    /nsm/zeek/logs/stats
57G     /nsm/zeek/logs/2022-03-04
42G     /nsm/zeek/logs/2022-03-05
43G     /nsm/zeek/logs/2022-03-06
40G     /nsm/zeek/logs/2022-03-07
900G    /nsm/zeek/logs

[username@heavynode2 ~]$ for PREFIX in $(ls -1 /nsm/zeek/logs/2022-03-02/ | cut -f 1 -d '.' | sort -u); do echo -n "$PREFIX: "; du -hc /nsm/zeek/logs/2022-03-02/${PREFIX}* | grep total | awk '{ print $1}'; done
broker: 96K
capture_loss: 192K
conn: 20G
conn-summary: 96K
dce_rpc: 319M
dhcp: 7.5M
dns: 7.1G
dpd: 11M
files: 1.4G
ftp: 13M
http: 3.6G
kerberos: 221M
known_certs: 104K
known_hosts: 208K
known_services: 3.7M
modbus: 1.0M
notice: 1.3M
ntlm: 27M
ntp: 250M
pe: 152K
radius: 5.0M
rdp: 1.4M
reporter: 92K
rfb: 28K
sip: 1.7M
smb_files: 296M
smb_mapping: 43M
smtp: 24M
snmp: 206M
socks: 100K
software: 936K
ssh: 19M
ssl: 521M
stats: 500K
syslog: 21G
tunnel: 552K
weird: 2.7G
x509: 291M
```

To summarize, the disk space is primarily occupied by `conn.*` and `syslog.*`, and somewhat by `dns.*`, `http.*`, `files.*`, and `weird.*`.

I'd also like to add - as far as I understand, we don't even _need_ these logs, as all the processed metadata gets stored in Elasticsearch anyway. I think a daily job to purge them would solve our disk space problem.

**Reply by dougburks** [reply]
There should be a cron job that runs `so-sensor-clean` every minute:
```
sudo crontab -l |grep so-sensor-clean
# SALT_CRON_IDENTIFIER:/usr/sbin/so-sensor-clean
* * * * * /usr/sbin/so-sensor-clean
```

`so-sensor-clean` should purge old Zeek logs when the disk reaches 90% usage. If your disk is at 90% now, you can manually delete old Zeek logs and then it should keep a smaller number of days worth of Zeek logs going forward.

Your capture_loss and weird logs are quite high. You'll want to look into those as that may be indicative of traffic problems from your upstream tap or span port.

**Reply by DrewHorrigan** [reply]
As I mentioned, `so-sensor-clean` deletes logs indiscriminately. I did purge the excessive Zeek logs at first, only leaving 3 days worth of logs, but they ended up accumulating anyway.

I ended up copying `/opt/so/saltstack/default/salt/zeek/init.sls` to `/opt/so/saltstack/local/salt/zeek/init.sls` and adding the following setting:
```
# Custom cron job to clean the zeek logs in accordance with zeekctl settings
zeekcleanlogscron:
  cron.present:
    - name: docker exec so-zeek /opt/zeek/bin/zeekctl cron
    - user: root
    - minute: '0'
    - hour: '0'
    - daymonth: '*'
    - month: '*'
    - dayweek: '*'
```

I am not sure it is the correct way to extend the state files, but my Saltstack skills are lacking. From what I understand, simply creating `/opt/so/saltstack/local/salt/zeek/init.sls` with the corresponding cron job would override the default `zeek/init.sls` file completely, which is not what I want (since it contains the code to run Zeek), but using the `extend` clause would require modifying some of the state files in "default" directory (since there are no files in `local/salt/zeek` that are loaded from the top file), in which case they would be overwritten in the event of an update. If someone has a better way to add a custom cron job on nodes running Zeek, please advise.

**Reply by dougburks** [reply]
If you really want to create your own cron job, then you can just do it manually without salt.

**Reply by DrewHorrigan** [reply]
That's what I ended up doing. I guess I'd prefer to configure everything from user-defined Salt files, but that works too.
